### ISSUE TITLE
Neo4j database crashes after running GDS-related functions, e.g. gds.louvain.mutate()

### ISSUE THREAD

**Issue by ThachNgocTran** [neo4j/graph-data-science#69]
## Description

I'm playing with Yelp Dataset. After projecting the original graph into `usersGraph` through:

```
CALL gds.graph.create(
    'usersGraph',
    'User',
    {
        FRIEND_OF: {
            type: 'FRIENDS',
            orientation: 'UNDIRECTED'
        }
    }
)
```

The projected graph is then fed into `louvain.mutate()` as followed:

```
CALL gds.louvain.mutate(
    'usersGraph',
    {
        mutateProperty: 'communityId'
    }
)
```

The graph result is saved into another Neo4j database:

```
CALL gds.graph.export('usersGraph', { dbName: 'tempdb' })
```

During the whole process, database `neo4j` could crash **arbitrarily**. The Neo4j Browser said that it lost the connection to the server. I moved back to Neo4j Desktop window, and tried to turn on the database again, but the database failed to start (A JRE popup appeared, asking to check the log files). I did try many times, e.g. even close Neo4j Desktop and open it again, or even restart the computer. The problem persisted.

I checked the log files (`neo4j.log`) and found this:

```
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  EXCEPTION_ACCESS_VIOLATION (0xc0000005) at pc=0x00007ffa07924323, pid=5224, tid=2224
#
# JRE version: OpenJDK Runtime Environment (Zulu11.35+13-CA) (11.0.5+10) (build 11.0.5+10-LTS)
# Java VM: OpenJDK 64-Bit Server VM (11.0.5+10-LTS, mixed mode, tiered, compressed oops, g1 gc, windows-amd64)
# Problematic frame:
# V  [jvm.dll+0x6a4323]
#
# No core dump will be written. Minidumps are not enabled by default on client versions of Windows
#
# An error report file with more information is saved as:
# C:\Users\[username]\AppData\Local\Programs\Neo4j Desktop\hs_err_pid5224.log
[thread 6396 also had an error]
#
# Compiler replay data is saved as:
# C:\Users\[username]\AppData\Local\Programs\Neo4j Desktop\replay_pid5224.log
#
# If you would like to submit a bug report, please visit:
#   http://www.azulsystems.com/support/
#
```

As a workaround, I created another Neo4j database with engine v4.0.6 (with APOC v4.0.0.17 and GDS v1.2.3 installed), everything worked smoothly. (created a projected graph, fed into Louvain, saved into another database, repeated the process many times).

## Reproducibility

+ Neo4j Desktop v1.3.2
+ Neo4j Database Engine v4.1.0 with setting `dbms.memory.heap.max_size=6G`.
+ Plugins: APOC v4.1.0, Graph Data Science Library v1.3.0
+ Windows 8.1 x64.
+ JDK 11.0.7 x64 (although Neo4j has built-in JRE, my computer has already JDK installed for other purposes. The `java.exe` can be reached through Command Prompt).

**Comment by ThachNgocTran** [comment]
I'm sorry. The Neo4j db actually crashed arbitrarily, not due to the lack of VC++ Redis 2017. I uninstalled this package, and still could create the projected graph, feed into louvain algorithm and even save the final graph into another database. But sometime the db just crashed.

I attached the "hs_err_pid5612.log" generated by Neo4j JVM (JRE version: OpenJDK Runtime Environment (Zulu11.35+13-CA)), for further investigation, and the neo4j log file.

[neo4j.log](https://github.com/neo4j/graph-data-science/files/4931849/neo4j.log)

[hs_err_pid5612.log](https://github.com/neo4j/graph-data-science/files/4931837/hs_err_pid5612.log)

**Comment by s1ck** [comment]
@ThachNgocTran Thanks for using GDS, reaching out and reporting that issue. I have a few questions:

Have you tried running your example on a different JRE than Azul, e.g. OpenJDK 11.0.6?
Does the example run fine without `graph.export` (there could be an UAC issue on Windows)?
APOC, doesn't seem to be necessary for your example, could you try disabling it?
Could you please try running with Neo4j v4.0.6 and GDS 1.3.1 to potentially rule out GDS?

A more general note .. Louvain is quite hungry for memory. For your particular use case Louvain consumes ~23GB (you can see the memory consumption in neo4j.log:4524). Increasing the heap size for Neo4j might be something to look into later.

Looking forward to your reply.

**Comment by ThachNgocTran** [comment]
@s1ck Hi, thanks for your response.

> Have you tried running your example on a different JRE than Azul, e.g. OpenJDK 11.0.6?

I haven't thought of this because I believe the default/built-in JRE for Neo4j should be already best suited.

**Just now, I have carried out a test, following your suggestions:**

> Does the example run fine without graph.export (there could be an UAC issue on Windows)?

"Neo4j Desktop" was run with "Run as Administrator" so that no permission issue could ever happen, even if UAC is turned on. You see the Integrity Level here.

![Started_With_High_Integrity](https://user-images.githubusercontent.com/9520123/89897700-b9d2a080-dbdf-11ea-89d5-40f09ccf286b.png)

> APOC, doesn't seem to be necessary for your example, could you try disabling it?

APOC was disabled.

> Could you please try running with Neo4j v4.0.6 and GDS 1.3.1 to potentially rule out GDS?

The latest 1.3.1 was downloaded [here](https://github.com/neo4j/graph-data-science/releases/tag/1.3.1). The JAR file was then renamed into "gdsLibrary-1.3.0.jar" and put in `\installation-4.1.0\plugins\` folder (overwrite the current one). I hope this is the correct way to install the latest version.

**However, the problem still persists**. It crashes randomly: sometime after `gds.graph.create`; sometime after `gds.graph.export`; sometime after running all those commands successfully, the database was _manually_ stopped in Neo4j Desktop and _manually_ started again. Those are the error banners:

![Error1](https://user-images.githubusercontent.com/9520123/89897407-429d0c80-dbdf-11ea-9706-684fca6283c2.png)
![Error2](https://user-images.githubusercontent.com/9520123/89897411-4335a300-dbdf-11ea-99ac-c7b4ce7b5be0.png)

After it crashed, the database would **no longer be usable**. Whenever _manually_ starting it again, I saw the banners above. Said in other words, the database was completely corrupted. The only way to fix is to delete it, create a new fresh/empty one and run my script again in order to have Yelp Database.

Please find attached the relevant log files:
[debug.log](https://github.com/neo4j/graph-data-science/files/5056959/debug.log)
[hs_err_pid3144.log](https://github.com/neo4j/graph-data-science/files/5056960/hs_err_pid3144.log)
[neo4j.log](https://github.com/neo4j/graph-data-science/files/5056961/neo4j.log)

**Comment by ThachNgocTran** [comment]
@s1ck Hi, I have further tested this issue with the latest Neo4j Desktop 1.3.4 (Graph Engine 4.1.1, GDS 1.3.1, without APOC) and Neo4j setting: `dbms.memory.heap.max_size=10G`).

![Neo4j Crash](https://user-images.githubusercontent.com/9520123/89931436-6aef3000-dc0c-11ea-94aa-324ad7226c82.png)

If memory requirement had been an issue (as you suggested: 23GB, but my computer has 16GB RAM and Graph Engine is allowed to use max 10GB), the command `gds.louvain.mutate` would have not returned.

**Comment by jjaderberg** [comment]
@ThachNgocTran Could you also try running outside of Neo4j Desktop? You can use the terminal in Desktop to navigate to the directory and start Neo4j with `bin\neo4j start`.

**Comment by ThachNgocTran** [comment]
Hi @jjaderberg 

Although I didn't try your approach (invoke `bin\neo4j.bat start`), I just tested and this issue is **no longer reproducible** with the **latest versions** of... everything:

+ Neo4j Desktop v1.3.4
+ Engine v4.1.2
+ APOC v4.1.0.2 (enabled)
+ Graph Data Science Library v1.3.3 (enabled)

Settings are default except `dbms.memory.heap.max_size=10G`.

After making the Yelp Dataset into Neo4j, I stopped the database, backed up the folder `installation-4.1.2`. Then carried out the 3 Cypher scripts above. Did not see any errors. Then stopped the database (via Stop button). Deleted the folder `installation-4.1.2`, and restored the original one. Again, ran the 3 scripts. Did not see any errors. I also tried to execute some more trivial Cyphers as well as turned off/on the database again. So far, the issue is no longer reproducible. :relaxed:
